File: mobile/plugins/kalm-vault/ios/Sources/KalmVaultPlugin/KalmVaultPlugin.swift
import AuthenticationServices
import Capacitor
import Foundation
import KalmVaultCore
import LocalAuthentication

/**
 * Unlocking the vault with Face ID or Touch ID, and the copy of the vault that
 * the AutoFill extension reads.
 *
 * The app hands this plugin a random 32-byte secret that the server knows as a
 * passkey. It goes into the Keychain behind biometrics, readable only on this
 * device and only while the same faces or fingers are enrolled. The web app
 * turns it back into the vault key; nothing here can open the vault on its own
 * and nothing here ever sees the master password.
 */
@objc(KalmVaultPlugin)
public class KalmVaultPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "KalmVaultPlugin"
    public let jsName = "KalmVault"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "status", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "enable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unlock", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveVault", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "autofillStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAutofillSettings", returnType: CAPPluginReturnPromise)
    ]

    private let store = SharedVaultStore()

    @objc func status(_ call: CAPPluginCall) {
        let context = LAContext()
        var error: NSError?
        var biometry = "none"
        if context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) {
            switch context.biometryType {
            case .faceID: biometry = "faceId"
            case .touchID: biometry = "touchId"
            default: biometry = "biometric"
            }
            if #available(iOS 17.0, *), context.biometryType == .opticID {
                biometry = "opticId"
            }
        }

        var result: [String: Any] = ["biometry": biometry, "enabled": false]
        if let meta = store.metadata() {
            result["enabled"] = true
            result["passkeyId"] = meta.passkeyId
        }
        call.resolve(result)
    }

    @objc func enable(_ call: CAPPluginCall) {
        guard
            let credentialId = call.getString("credentialId"),
            let passkeyId = call.getString("passkeyId"),
            let secret = call.getString("secret"),
            let wrappedKey = call.getString("wrappedKey")
        else {
            call.reject("Missing credential details.", "invalid")
            return
        }

        // Asked first, so turning this on is a deliberate act with the face or
        // finger that will later unlock the vault.
        let context = LAContext()
        context.evaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            localizedReason: "Turn on unlocking KalmPass with biometrics"
        ) { success, error in
            guard success else {
                Self.rejectForAuthentication(call, error)
                return
            }
            do {
                try self.store.saveSecret(
                    UnlockSecret(credentialId: credentialId, secret: secret),
                    context: context
                )
                try self.store.saveMetadata(
                    UnlockMetadata(credentialId: credentialId, passkeyId: passkeyId, wrappedKey: wrappedKey)
                )
                call.resolve()
            } catch {
                self.store.clear()
                call.reject("Could not store the unlock key: \(error.localizedDescription)", "storage")
            }
        }
    }

    @objc func unlock(_ call: CAPPluginCall) {
        let reason = call.getString("reason") ?? "Unlock your vault"
        DispatchQueue.global(qos: .userInitiated).async {
            switch self.store.readSecret(reason: reason) {
            case .success(let unlock):
                call.resolve(["credentialId": unlock.credentialId, "secret": unlock.secret])
            case .failure(.cancelled):
                call.reject("Cancelled.", "cancelled")
            case .failure(.invalidated):
                // Biometrics changed since this was set up. The item can never
                // be read again, so it goes.
                self.store.clear()
                call.reject(
                    "Your phone's biometrics changed. Unlock with your master password and turn this on again.",
                    "invalidated"
                )
            case .failure(.other(let status)):
                call.reject("Could not read the unlock key (\(status)).", "storage")
            }
        }
    }

    @objc func disable(_ call: CAPPluginCall) {
        store.clear()
        ASCredentialIdentityStore.shared.removeAllCredentialIdentities(nil)
        call.resolve()
    }

    /**
     * The vault as the server holds it, ciphertext only, for the AutoFill
     * extension. The site and username of each login also go to the system's
     * credential identity store, so iOS can suggest them above the keyboard.
     */
    @objc func saveVault(_ call: CAPPluginCall) {
        guard store.metadata() != nil else {
            call.resolve()
            return
        }

        let items: [VaultItem] = (call.getArray("items") ?? []).compactMap { entry in
            guard
                let row = entry as? [String: Any],
                let id = row["id"] as? String,
                let data = row["data"] as? String
            else { return nil }
            return VaultItem(id: id, data: data)
        }

        do {
            try store.saveItems(items)
        } catch {
            call.reject("Could not save the vault for AutoFill: \(error.localizedDescription)", "storage")
            return
        }

        let identities: [ASPasswordCredentialIdentity] = (call.getArray("identities") ?? []).compactMap { entry in
            guard
                let row = entry as? [String: Any],
                let id = row["id"] as? String,
                let url = row["url"] as? String,
                let username = row["username"] as? String,
                let host = VaultMatch.host(of: url)
            else { return nil }
            return ASPasswordCredentialIdentity(
                serviceIdentifier: ASCredentialServiceIdentifier(identifier: host, type: .domain),
                user: username,
                recordIdentifier: id
            )
        }

        ASCredentialIdentityStore.shared.getState { state in
            guard state.isEnabled else {
                call.resolve()
                return
            }
            ASCredentialIdentityStore.shared.replaceCredentialIdentities(with: identities) { _, _ in
                call.resolve()
            }
        }
    }

    @objc func autofillStatus(_ call: CAPPluginCall) {
        ASCredentialIdentityStore.shared.getState { state in
            call.resolve(["supported": true, "enabled": state.isEnabled])
        }
    }

    /** iOS 17 can open the right page of Settings. Earlier versions cannot. */
    @objc func openAutofillSettings(_ call: CAPPluginCall) {
        if #available(iOS 17.0, *) {
            ASSettingsHelper.openCredentialProviderAppSettings { error in
                if let error {
                    call.reject(error.localizedDescription, "unavailable")
                } else {
                    call.resolve(["opened": true])
                }
            }
        } else {
            call.resolve(["opened": false])
        }
    }

    private static func rejectForAuthentication(_ call: CAPPluginCall, _ error: Error?) {
        let code = (error as? LAError)?.code
        let cancelled: [LAError.Code] = [.userCancel, .appCancel, .systemCancel, .userFallback]
        if let code, cancelled.contains(code) {
            call.reject("Cancelled.", "cancelled")
        } else {
            call.reject(error?.localizedDescription ?? "Biometrics are not available.", "unavailable")
        }
    }
}

File: mobile/plugins/kalm-vault/android/src/main/java/net/kalmpass/vault/KalmVaultPlugin.java
package net.kalmpass.vault;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.security.keystore.KeyPermanentlyInvalidatedException;
import android.view.autofill.AutofillManager;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import javax.crypto.Cipher;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Unlocking the vault with a fingerprint or face, and the copy of the vault
 * the autofill service reads.
 *
 * The app hands this plugin a random 32-byte secret that the server knows as a
 * passkey. It is kept encrypted under a Keystore key that needs a strong
 * biometric check for every use. Nothing here can open the vault by itself,
 * and nothing here ever sees the master password.
 */
@CapacitorPlugin(name = "KalmVault")
public class KalmVaultPlugin extends Plugin {

    private UnlockStore store() {
        return new UnlockStore(getContext());
    }

    @PluginMethod
    public void status(PluginCall call) {
        UnlockStore store = store();
        JSObject result = new JSObject();
        result.put("biometry", BiometricGate.available(getContext()) ? "biometric" : "none");
        result.put("enabled", store.isEnabled());
        if (store.isEnabled()) result.put("passkeyId", store.passkeyId());
        call.resolve(result);
    }

    @PluginMethod
    public void enable(PluginCall call) {
        String credentialId = call.getString("credentialId");
        String passkeyId = call.getString("passkeyId");
        String secret = call.getString("secret");
        String wrappedKey = call.getString("wrappedKey");
        if (credentialId == null || passkeyId == null || secret == null || wrappedKey == null) {
            call.reject("Missing credential details.", "invalid");
            return;
        }

        UnlockStore store = store();
        Cipher cipher;
        try {
            cipher = store.cipherForSaving();
        } catch (Exception error) {
            store.clear();
            call.reject("This phone cannot hold an unlock key: " + error.getMessage(), "unavailable");
            return;
        }

        BiometricGate.ask(
            getActivity(),
            "Unlock KalmPass with biometrics",
            "Confirm it is you to turn this on",
            cipher,
            new BiometricGate.Callback() {
                @Override
                public void onSuccess(Cipher authenticated) {
                    try {
                        store.save(authenticated, credentialId, passkeyId, secret, wrappedKey);
                        call.resolve();
                    } catch (Exception error) {
                        store.clear();
                        call.reject("Could not store the unlock key: " + error.getMessage(), "storage");
                    }
                }

                @Override
                public void onCancel() {
                    store.clear();
                    call.reject("Cancelled.", "cancelled");
                }

                @Override
                public void onError(String message) {
                    store.clear();
                    call.reject(message, "unavailable");
                }
            }
        );
    }

    @PluginMethod
    public void unlock(PluginCall call) {
        UnlockStore store = store();
        Cipher cipher;
        try {
            cipher = store.cipherForReading();
        } catch (KeyPermanentlyInvalidatedException changed) {
            // New biometrics were enrolled. The key is gone for good.
            store.clear();
            call.reject(
                "Your phone's biometrics changed. Unlock with your master password and turn this on again.",
                "invalidated"
            );
            return;
        } catch (Exception error) {
            call.reject("Could not read the unlock key: " + error.getMessage(), "storage");
            return;
        }

        String reason = call.getString("reason", "Unlock your vault");
        BiometricGate.ask(
            getActivity(),
            "KalmPass",
            reason,
            cipher,
            new BiometricGate.Callback() {
                @Override
                public void onSuccess(Cipher authenticated) {
                    try {
                        JSONObject unlock = store.open(authenticated);
                        JSObject result = new JSObject();
                        result.put("credentialId", unlock.getString("credentialId"));
                        result.put("secret", unlock.getString("secret"));
                        call.resolve(result);
                    } catch (Exception error) {
                        call.reject("Could not read the unlock key: " + error.getMessage(), "storage");
                    }
                }

                @Override
                public void onCancel() {
                    call.reject("Cancelled.", "cancelled");
                }

                @Override
                public void onError(String message) {
                    call.reject(message, "unavailable");
                }
            }
        );
    }

    @PluginMethod
    public void disable(PluginCall call) {
        store().clear();
        call.resolve();
    }

    /** Ciphertext only, as the server holds it, for the autofill service. */
    @PluginMethod
    public void saveVault(PluginCall call) {
        UnlockStore store = store();
        if (!store.isEnabled()) {
            call.resolve();
            return;
        }
        try {
            JSArray items = call.getArray("items", new JSArray());
            JSONArray rows = new JSONArray();
            for (int i = 0; i < items.length(); i++) {
                JSONObject item = items.getJSONObject(i);
                rows.put(new JSONObject().put("id", item.getString("id")).put("data", item.getString("data")));
            }
            store.saveItems(rows.toString());
            call.resolve();
        } catch (Exception error) {
            call.reject("Could not save the vault for autofill: " + error.getMessage(), "storage");
        }
    }

    @PluginMethod
    public void autofillStatus(PluginCall call) {
        JSObject result = new JSObject();
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            result.put("supported", false);
            result.put("enabled", false);
        } else {
            AutofillManager manager = getContext().getSystemService(AutofillManager.class);
            result.put("supported", manager != null && manager.isAutofillSupported());
            result.put("enabled", manager != null && manager.hasEnabledAutofillServices());
        }
        call.resolve(result);
    }

    @PluginMethod
    public void openAutofillSettings(PluginCall call) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            call.resolve(new JSObject().put("opened", false));
            return;
        }
        Intent intent = new Intent(Settings.ACTION_REQUEST_SET_AUTOFILL_SERVICE);
        intent.setData(Uri.parse("package:" + getContext().getPackageName()));
        try {
            getActivity().startActivity(intent);
            call.resolve(new JSObject().put("opened", true));
        } catch (Exception error) {
            call.resolve(new JSObject().put("opened", false));
        }
    }
}
